components/waypoint-movement.js

```diff
@@ -0,0 +1,239 @@
+      this.positions = new Array(this.children.length);
+      for (var i = 0; i < this.children.length; i++) {
+        this.positions[i] = new Float32Array(3);
+        this.children[i].getTranslationWorld(this.positions[i]);
+      }
+      this.currentPositionIndex = 0;
+      this.currentPosition = new Float32Array(3);
+      this.lookPosition = new Float32Array(3);
+      this.fromPosition = this.positions[0];
+      this.toPosition = this.positions[1];
+      glMatrix.vec3.sub(this.currentPosition, this.toPosition, this.fromPosition);
+      this.fromTolength = glMatrix.vec3.length(this.currentPosition);
+      this.currentLength = 0;
+      this.up = [0, 1, 0];
+      this.quat = new Float32Array(4);
+      lookAt(this.quat, this.toPosition, this.fromPosition, this.up);
+      this.object.resetTranslationRotation();
+      this.object.rotate(this.quat);
+      this.object.setTranslationWorld(this.fromPosition);
+
+      this.p0 = new Float32Array(3);
+      this.p1 = new Float32Array(3);
+      this.p2 = new Float32Array(3);
+      this.currentCurveIndex = 0;
+      this.curveDistance = Math.max(0.0, Math.min(0.49999, this.curveDistance));
+      this.bezFactorMultiplicator = 1 / (2*this.curveDistance);
+      this._setCurvePoints();
+
+      this.onFinalWaypointReachedCallbacks = [];
+    },
+    /**
+     * Adds a function to a callback array which, then calls the function after the final waypoint has been reached.
+     * @param  {[type]} f Function to be added.
+     */
+    addOnFinalWaypointReachedCallback: function(f) {
+      this.onFinalWaypointReachedCallbacks.push(f);
+    },
+    /**
+     * Removes a function from the callback array for reaching the final waypoint..
+     * @param  {[type]} f Function to be removed.
+     */
+    removeOnFinalWaypointReachedCallback: function(f) {
+      const index = array.indexOf(f);
+      if (index > -1) {
+        this.onFinalWaypointReachedCallbacks.splice(index, 1);
+      }
+    },
+    /**
+     * Sets the points needed for the next curve
+     */
+     _setCurvePoints: function(){
+      glMatrix.vec3.sub(this.currentPosition, this.positions[this.currentCurveIndex], this.positions[this.currentCurveIndex + 1])
+      glMatrix.vec3.scale(this.currentPosition, this.currentPosition, this.curveDistance);
+      glMatrix.vec3.add(this.p0, this.positions[this.currentCurveIndex + 1], this.currentPosition);
+
+      this.p1 = this.positions[this.currentCurveIndex + 1];
+
+      glMatrix.vec3.sub(this.currentPosition, this.positions[this.currentCurveIndex + 2], this.positions[this.currentCurveIndex + 1])
+      glMatrix.vec3.scale(this.currentPosition, this.currentPosition, this.curveDistance);
+      glMatrix.vec3.add(this.p2, this.positions[this.currentCurveIndex + 1], this.currentPosition);
+    },
+    /**
+     * Sets the object's path object and configures it.
+     * @param  {[WL.Object]} pathObject [Object containing alphebetically named children]
```

Comment:
From the JSDoc here, the function expects an array of objects. You probably inteded `{WL.Object}` :)